### ISSUE TITLE
API Proposal for multiline quick input

### ISSUE THREAD

**Issue by devanshj** [microsoft/vscode#98098]
## Problem

Currently there's no way to take a multiline input via `showInputBox`. We need one because we use `showInputBox` to take the commit message (which can be multiline) in git extension. (#40295, #39969)

## Solution APIs

TLDR: allow passing a `multiline` property to the `showInputBox` options

- **PROPOSED** API: Allow passing a `multiline` property to the `showInputBox` options.

	- Problem: Users can have `{ multiline: true, password: true }` which won't work because it's a textarea. More importantly in future there could be more options applicable to only multiline (eg an option mapping to textarea's `row` attribute) or to only singleline

		- **PROPOSED** Solution: It's okay just document it
		- Solution: It's okay just document it and throw an error if such case arises
		- Solution: make the type of options like ```{ multiline?: false, password?: boolean } | { multiline?: true, password?: false }```

            - Problem: Terrible typescript inference and error message ergonomics if the options are not literal
			- Problem: It's typescript-only solutiton

- API: Have another function called `showMultilineInputBox` that takes options only applicable to it (meaning doesn't take `password` property as option)

	- Problem: Feels a little overkill especially at the moment as we only one option (ie `password`) exclusive to singleline. It would have been okay if we had a bunch of options only exclusive to multiline eg `maxHeight`, `autoGrow`, `initialRows`, etc. or bunch of options exclusive to singleline

## Solution Behaviors

TLDR: when `multiline` is set to true, `showInputBox` shows a multline quick input where the user hits shift+enter to insert a newline and works exactly like current quick input

- **PROPOSED** Behavior: Allow users to insert a newline with shift+enter instead of submitting
	- Prior art: In chrome developer console enter executes that code but shift+enter inserts a newline. As in, usually afaik in scenarios where enter is already taken, shift+enter mimics the "usual" enter behavior.
- Behavior: Allow users to insert a newline with enter instead of submitting and have shift+enter submit it.
	- Problem: The behavior is not "additive" to singleline behavior as user would expect enter to submit it in multiline also as it does in singleline
- Behavior: Allow users to insert a newline with enter instead of submitting and have a button to submit it. 
	- Problem: Kinda kills the whole point of "quick input" you'll have to press tab to focus to button and hit enter, foreign to quick action ui and not additive to singleline behavior 

## Implementation

For currently proposed api and behavior: #98042, [demo](https://github.com/microsoft/vscode/issues/40295#issuecomment-629862549)

cc @jrieken @chrmarti @joaomoreno

**Comment by devanshj** [comment]
Hey @chrmarti, this proposal fixes a bug in git extension that has 112 upvotes and is 3 years old, not to mention it's a small change which is no-brainer & is pretty straight-forward. So I guess the reward to investment ratio is high enough that it would be great if you can prioritize it a little thanks :)

**Comment by joaomoreno** [comment]
@devanshj Please have patience and don't ping us. As you can see there are 5k issues and 200 PRs open, on top of all the monthly releases. There's a lot to work on. We'll get to it in time.

**Comment by devanshj** [comment]
FWIW I wouldn't call it a "feature-request" (I'm not requesting a feature) it's more of a "bugfix" (I'm providing a bugfix that involves adding a feature) because it's [part of the process](https://github.com/microsoft/vscode/pull/98042#issuecomment-630010040) of getting [this PR](https://github.com/microsoft/vscode/pull/98042#issuecomment-630010040) merged.

To answer the bot, [#40295](https://github.com/microsoft/vscode/issues/40295) (the issue that gets fixed by this proposal & PR) already has 112 upvotes

**Comment by xenoterracide** [comment]
I was just going to open a ticket to simply make the git commit box multiline... how do I vote for this issue?

**Comment by devanshj** [comment]
> I was just going to open a ticket to simply make the git commit box multiline... how do I vote for this issue?

The 👍 you did counts as a vote afaict so you're good! Also this already has enough votes in the issues I've linked the reason I opened this it to discuss the api can track the lifecycle of this according to the guildlines.
So actually you should also 👍 [this](https://github.com/microsoft/vscode/issues/39969) which imo is the actual "feature-request".

**Comment by PELock** [comment]
Just started coding my extension for VSCode, it lacs basic UI input methods. No multiline inputs, no way to design a simple dialog box (we need to use Electron apps like the Color Palette is using and do IPC...). I'm surprised VScode extension API lacks so many basic features.

**Comment by nvuillam** [comment]
Feature needed, pleaaaase :)

**Comment by PELock** [comment]
2022 - still no multiline inputs hahah

**Comment by yi-Xu-0100** [comment]
Now is 2023, and still wait for this feature...😶

By the way, the ChatGPT suggestion use `input.multiline = true` to realise the multiline inputs. hahaha🤣

**Comment by PELock** [comment]
Yeah, nobody from MS gives a flying F***

**Comment by irataxy** [comment]
Any update on this?

**Comment by PELock** [comment]
Yeah BIG FU

**Comment by JamyDev** [comment]
@isidorn is this something that can be brought up for consideration? Looks like @eamodio already has a PR ready with an implementation.

**Comment by erlangparasu** [comment]
Any update?

**Comment by pwsses** [comment]
Any Update??

**Comment by emorrisn** [comment]
Any update?

**Comment by MarkShawn2020** [comment]
Any update?

**Comment by PELock** [comment]
Yeah the update title is if fck you :D
